src/ai/ai-bedrock-cohere-embedding.ts

```diff
@@ -0,0 +1,155 @@
+} from './bedrock-embedding-service-schemas.js';
+
+if (!process.env.AWS_REGION) {
+  throw new Error('Please set AWS_REGION');
+}
+if (!process.env.AWS_ACCESS_KEY_ID) {
+  throw new Error('Please set AWS_ACCESS_KEY_ID');
+}
+if (!process.env.AWS_SECRET_ACCESS_KEY) {
+  throw new Error('Please set AWS_SECRET_ACCESS_KEY');
+}
+
+const bedrockClient = new BedrockRuntimeClient({
+  region: process.env.AWS_REGION,
+  credentials: {
+    accessKeyId: process.env.AWS_ACCESS_KEY_ID as string,
+    secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY as string,
+  },
+});
+
+/**
+ * Verify that the string sent has less tokens than the maximum possible for the model
+ * @param root named parameters
+ * @param root.chunk the chunk to verify
+ * @returns true or false
+ */
+function exceedsTokenLimit({ chunk }: { chunk: string }): boolean {
+  const encoding = new Tiktoken(
+    cl100k.bpe_ranks,
+    cl100k.special_tokens,
+    cl100k.pat_str,
+  );
+  const tokens = encoding.encode(chunk);
+  encoding.free();
+  // cohere input token limit = 512
+  return tokens.length > 512;
+}
+
+/**
+ * Call the bedrock embedding model
+ * @param root named parameters
+ * @param root.chunks A document, or a list of chunks from a document to embed
+ * @param root.isQuery is the documnt for query
+ * @returns an object with the averaged embedding an an array of chunk embeddings
+ */
+export async function invokeBedrockCohereEmbedding({
+  chunks,
+  isQuery,
+}: {
+  chunks: string[];
+  isQuery: boolean;
+}): Promise<CohereEmbeddingResponse[] | undefined> {
+  const processedChunks = await Promise.all(
+    chunks.map(async (chunk) => {
+      return chunk.replace('\n', '');
+    }),
+  );
+  try {
+    if (
+      chunks.some((chunk) => {
+        return exceedsTokenLimit({ chunk });
+      })
+    ) {
+      throw new Error('A document exceeds token limit');
+    }
+    const embeddingResponses = Promise.all(
+      processedChunks.map(async (chunk) => {
+        const command = new InvokeModelCommand({
+          modelId: 'cohere.embed-english-v3',
+          body: JSON.stringify({
+            texts: [chunk],
+            input_type: isQuery ? 'search_query' : 'search_document',
+          } satisfies CohereEmbeddingBody),
+        });
+        const response = await bedrockClient.send(command);
+        const responseBody: CohereEmbeddingResponse = JSON.parse(
+          Buffer.from(response.body).toString('utf8'),
+        );
+        return responseBody;
+      }),
+    );
+    return embeddingResponses;
+  } catch (error) {
+    const message = 'The embedding function had an error';
+    console.error({
+      error,
+      message,
+    });
+    return undefined;
+  }
+}
+
+/**
+ * Split a document into chunks. This is a simple way of creating chunks and not context-aware
+ * @param root named parameters
+ * @param root.document the document to split into chunks
+ * @returns a list of chunks
+ */
+export async function createDocumentChunks({
```

Comment:
Good improvement over my implementation of this function (I did not plan for filtering); However you still need to remove `\n` before the splitter (unless there is a reason to keep them of course)

Additionally, remove duplicate spaces, this can happen at times
```
// trim and remove duplicate whitespace
const trimmedDocument = cleanedDocument.replaceAll(/\s+/g, ' ').trim();
```